qlib/Swagger.qm

```diff
@@ -0,0 +1,3666 @@
+%endif
+        }
+        return new SwaggerSchema(val);
+    }
+
+    #! Load <a href="https://swagger.io/">Swagger</a> definition from a file.
+    /**
+        @param filepath path to the <a href="https://swagger.io/">Swagger</a> file
+
+        @throws EMPTY-SCHEMA-FILE %Swagger file is empty
+        @throws JSON-MODULE-MISSING trying to parse a JSON %Swagger specification with the json module unavailable
+        @throws YAML-MODULE-MISSING trying to parse a YAML %Swagger specification with the yaml module unavailable
+        @throws XML-MODULE-MISSING trying to parse an XML %Swagger specification with the xml module unavailable
+     */
+    static SwaggerSchema fromFile(string filepath) {
+        return new SwaggerSchema(SwaggerLoader::parseSchemaContent(filepath, ReadOnlyFile::readTextFile(filepath)));
+    }
+
+    #! Load a schema definition from a file
+    /**
+        @param filepath path to the schema file
+        @param str the file content to parse
+
+        @throws EMPTY-SCHEMA-FILE schema file is empty
+        @throws JSON-MODULE-MISSING trying to parse a JSON schema with the json module unavailable
+        @throws YAML-MODULE-MISSING trying to parse a YAML schema specification with the yaml module unavailable
+        @throws XML-MODULE-MISSING trying to parse an XML schema specification with the xml module unavailable
+     */
+    static hash parseSchemaContent(string filepath, string str) {
+        if (!str)
+            throw "EMPTY-SCHEMA-FILE", sprintf("%y: file is empty", filepath);
+
+        string ser = "json";
+        string filename = basename(filepath);
+        if (filename =~ /.*\.json$/i) {
+            ser = "json";
+        }
+        else if (filename =~ /.*\.yaml$/i) {
+            ser = "yaml";
+        }
+        else if (filename =~ /.*\.xml$/i) {
+            ser = "xml";
+        }
+        else {
+            DataLineIterator it(str);
+            while (it.next()) {
+                string line = trim(it.getValue());
+                if (line.size()) {
+                    if (line =~ /^<\?xml/)
+                        ser = "xml";
+                    else if (line =~ /^([^{])+:/ || line =~ /^(---|%YAML)/)
+                        ser = "yaml";
+                    else
+                        ser = "json";
+                    break;
+                }
+            }
+        }
+
+        hash rv;
+        switch (ser) {
+            case "json": {
+%ifdef NoJson
+                throw "JSON-MODULE-MISSING", "Trying to parse a JSON schema, but the json module is unavailable";
+%else
+                rv = parse_json(str);
+                break;
+%endif
+            }
+            case "yaml": {
+%ifdef NoYaml
+                throw "YAML-MODULE-MISSING", "Trying to parse a YAML schema, but the yaml module is unavailable";
+%else
+                rv = parse_yaml(str);
+                break;
+%endif
+            }
+            case "xml": {
```

Comment:
Why is there an option for XML when Swagger specifications are only JSON or YAML?

Reply:
Swagger 2.0 specifies XML de/serialization, however it has not yet been implemented yet